# SP21-DATA601-HW-4

Due Date: 03/09/2021 (Tuesday) by 23:59 pm 

# Part-I: Set-up

1. Make sure that you make a copy of this notebook.

2. Run the cell below.

In [271]:
## Don't change this cell and make sure that you run this cell.
!wget https://github.com/mguner/DATA601_SP21/raw/main/data/Archive.zip

--2021-03-10 01:43:25--  https://github.com/mguner/DATA601_SP21/raw/main/data/Archive.zip
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/mguner/DATA601_SP21/main/data/Archive.zip [following]
--2021-03-10 01:43:25--  https://raw.githubusercontent.com/mguner/DATA601_SP21/main/data/Archive.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 104858 (102K) [application/zip]
Saving to: ‘Archive.zip.3’

Archive.zip.3       100%[===================>] 102.40K  --.-KB/s    in 0.008s  

2021-03-10 01:43:25 (13.1 MB/s) - ‘Archive.zip.3’ saved [104858/104858]



3. Unzip the file

There should be a zipped file in `/content/Archive.zip` unzip this file in `content` folder. (Note: `content` directory is the current working directory. So you don't need to do anything extra)

__Note:__ If you cannot solve this problem then you will not be able to work on this homework. In this case, you can buy the answer by reaching out to me as an exchange of 20% of your grade. 

In [272]:
## your answer is here
from zipfile import ZipFile

In [273]:
with ZipFile('/content/Archive.zip','r') as zip:
   zip.printdir()
   zip.extractall(path='content')

File Name                                             Modified             Size
dmv_hs.csv                                     2021-03-03 21:20:00       482313
__MACOSX/._dmv_hs.csv                          2021-03-03 21:20:00          368
maryland_hs.csv                                2021-03-03 20:50:16        25851
__MACOSX/._maryland_hs.csv                     2021-03-03 20:50:16          312


4. Import pandas and read the datasets

In [274]:
## Assuming you were able to unzip the file.
## Don't change the names of the datasets.
import pandas as pd
maryland_hs =pd.read_csv('/content/content/maryland_hs.csv',index_col= 0)
dmv_hs = pd.read_csv('/content/content/dmv_hs.csv',index_col= 0)

In [275]:
dmv_hs

school_name  ... state_school_id
0              A G RICHARDSON ELEM  ...  VA-024-0240240
1                   A M DAVIS ELEM  ...  VA-021-0210620
2                A. HENDERSON ELEM  ...  VA-075-0750330
3      A. LINWOOD HOLTON GOV. SCH.  ...  VA-268-2680010
4       A. MARIO LOIEDERMAN MIDDLE  ...    MD-15-150787
...                            ...  ...             ...
3763       YOUGH GLADES ELEMENTARY  ...    MD-11-110710
3764     YOUTHS BENEFIT ELEMENTARY  ...    MD-12-120348
3765                   YOWELL ELEM  ...  VA-024-0240040
3766                     YUMA ELEM  ...  VA-084-0840790
3767  YVONNE B. MILLER HIGH SCHOOL  ...  VA-917-9170101

[3768 rows x 12 columns]

In [276]:
maryland_hs

School Name  ... Urban-centric Locale [Public School] 2018-19
0                             ABERDEEN HIGH  ...                          22-Suburb: Mid-size
1        ACADEMY OF HEALTH SCIENCES AT PGCC  ...                             21-Suburb: Large
2    ACHIEVEMENT ACADEMY @ HARBOR CITY HIGH  ...                               11-City: Large
3                      ALBERT EINSTEIN HIGH  ...                             21-Suburb: Large
4                             ALLEGANY HIGH  ...                               13-City: Small
..                                      ...  ...                                          ...
199                         WILDE LAKE HIGH  ...                               13-City: Small
200                       WILLIAMSPORT HIGH  ...                          22-Suburb: Mid-size
201                  WINSTON CHURCHILL HIGH  ...                             21-Suburb: Large
202                       WINTERS MILL HIGH  ...                             41-Rural: Fringe
203  WOODLAWN HIGH CENTER FOR PRE-ENG. RES.  ...                             21-Suburb: Large

[204 rows x 13 columns]

# Part-II: Questions

## Problem

Rename columns in the [snake case](https://en.wikipedia.org/wiki/Snake_case#:~:text=Snake%20case%20(stylized%20as%20snake_case,subroutine%20names%2C%20and%20for%20filenames.) format.

In [277]:
## your answer is here
dmv_hs.columns = dmv_hs.columns.str.replace(' ','_')
dmv_hs.columns = dmv_hs.columns.str.lower()
maryland_hs.columns = maryland_hs.columns.str.replace(' ','_')
maryland_hs.columns = maryland_hs.columns.str.lower()

In [278]:
dmv_hs

school_name  ... state_school_id
0              A G RICHARDSON ELEM  ...  VA-024-0240240
1                   A M DAVIS ELEM  ...  VA-021-0210620
2                A. HENDERSON ELEM  ...  VA-075-0750330
3      A. LINWOOD HOLTON GOV. SCH.  ...  VA-268-2680010
4       A. MARIO LOIEDERMAN MIDDLE  ...    MD-15-150787
...                            ...  ...             ...
3763       YOUGH GLADES ELEMENTARY  ...    MD-11-110710
3764     YOUTHS BENEFIT ELEMENTARY  ...    MD-12-120348
3765                   YOWELL ELEM  ...  VA-024-0240040
3766                     YUMA ELEM  ...  VA-084-0840790
3767  YVONNE B. MILLER HIGH SCHOOL  ...  VA-917-9170101

[3768 rows x 12 columns]

In [279]:
maryland_hs

school_name  ... urban-centric_locale_[public_school]_2018-19
0                             ABERDEEN HIGH  ...                          22-Suburb: Mid-size
1        ACADEMY OF HEALTH SCIENCES AT PGCC  ...                             21-Suburb: Large
2    ACHIEVEMENT ACADEMY @ HARBOR CITY HIGH  ...                               11-City: Large
3                      ALBERT EINSTEIN HIGH  ...                             21-Suburb: Large
4                             ALLEGANY HIGH  ...                               13-City: Small
..                                      ...  ...                                          ...
199                         WILDE LAKE HIGH  ...                               13-City: Small
200                       WILLIAMSPORT HIGH  ...                          22-Suburb: Mid-size
201                  WINSTON CHURCHILL HIGH  ...                             21-Suburb: Large
202                       WINTERS MILL HIGH  ...                             41-Rural: Fringe
203  WOODLAWN HIGH CENTER FOR PRE-ENG. RES.  ...                             21-Suburb: Large

[204 rows x 13 columns]

## Problem: Extra

Can you get rid of the strings "[public_school]" and "2018-19" and "(excludes_ae)" from column names?

In [280]:
## your answer is here
new_columns=[]
for column in maryland_hs.columns:
  column=column.replace('_[public_school]','')
  column=column.replace('_2018-19','')
  column=column.replace('_(excludes_ae)','')
  new_columns.append(column)
maryland_hs.columns = new_columns  
maryland_hs

school_name  ... urban-centric_locale
0                             ABERDEEN HIGH  ...  22-Suburb: Mid-size
1        ACADEMY OF HEALTH SCIENCES AT PGCC  ...     21-Suburb: Large
2    ACHIEVEMENT ACADEMY @ HARBOR CITY HIGH  ...       11-City: Large
3                      ALBERT EINSTEIN HIGH  ...     21-Suburb: Large
4                             ALLEGANY HIGH  ...       13-City: Small
..                                      ...  ...                  ...
199                         WILDE LAKE HIGH  ...       13-City: Small
200                       WILLIAMSPORT HIGH  ...  22-Suburb: Mid-size
201                  WINSTON CHURCHILL HIGH  ...     21-Suburb: Large
202                       WINTERS MILL HIGH  ...     41-Rural: Fringe
203  WOODLAWN HIGH CENTER FOR PRE-ENG. RES.  ...     21-Suburb: Large

[204 rows x 13 columns]

In [281]:
## Don't change anything in this cell
## Make sure you run this cell

new_columns = ['school_name',
 'state_name_latest_available_year',
 'lowest_grade_offered',
 'highest_grade_offered',
 'total_students_all_grades',
 'free_and_reduced_lunch_students',
 'reduced-price_lunch_eligible_students',
 'latitude',
 'longitude',
 'magnet_school',
 'charter_school',
 'shared_time_school',
 'urban-centric_locale']
maryland_hs.columns = new_columns
maryland_hs.head(2)

school_name  ... urban-centric_locale
0                       ABERDEEN HIGH  ...  22-Suburb: Mid-size
1  ACADEMY OF HEALTH SCIENCES AT PGCC  ...     21-Suburb: Large

[2 rows x 13 columns]

## Problem 

In both `maryland_hs` and `dmv_hs` rename the column `state_name_latest_available_year` with `state_name`. Make sure that the changes reflected in the original dataset.

In [282]:
## your answer is here
maryland_hs.rename(columns={'state_name_latest_available_year':'state_name'},inplace=True)
dmv_hs.rename(columns={'state_name_latest_available_year':'state_name'},inplace=True)

In [283]:
maryland_hs

school_name  ... urban-centric_locale
0                             ABERDEEN HIGH  ...  22-Suburb: Mid-size
1        ACADEMY OF HEALTH SCIENCES AT PGCC  ...     21-Suburb: Large
2    ACHIEVEMENT ACADEMY @ HARBOR CITY HIGH  ...       11-City: Large
3                      ALBERT EINSTEIN HIGH  ...     21-Suburb: Large
4                             ALLEGANY HIGH  ...       13-City: Small
..                                      ...  ...                  ...
199                         WILDE LAKE HIGH  ...       13-City: Small
200                       WILLIAMSPORT HIGH  ...  22-Suburb: Mid-size
201                  WINSTON CHURCHILL HIGH  ...     21-Suburb: Large
202                       WINTERS MILL HIGH  ...     41-Rural: Fringe
203  WOODLAWN HIGH CENTER FOR PRE-ENG. RES.  ...     21-Suburb: Large

[204 rows x 13 columns]

In [284]:
dmv_hs

school_name state_name  ...  state_agency_id state_school_id
0              A G RICHARDSON ELEM   Virginia  ...           VA-024  VA-024-0240240
1                   A M DAVIS ELEM   Virginia  ...           VA-021  VA-021-0210620
2                A. HENDERSON ELEM   Virginia  ...           VA-075  VA-075-0750330
3      A. LINWOOD HOLTON GOV. SCH.   Virginia  ...           VA-268  VA-268-2680010
4       A. MARIO LOIEDERMAN MIDDLE   Maryland  ...            MD-15    MD-15-150787
...                            ...        ...  ...              ...             ...
3763       YOUGH GLADES ELEMENTARY   Maryland  ...            MD-11    MD-11-110710
3764     YOUTHS BENEFIT ELEMENTARY   Maryland  ...            MD-12    MD-12-120348
3765                   YOWELL ELEM   Virginia  ...           VA-024  VA-024-0240040
3766                     YUMA ELEM   Virginia  ...           VA-084  VA-084-0840790
3767  YVONNE B. MILLER HIGH SCHOOL   Virginia  ...           VA-917  VA-917-9170101

[3768 rows x 12 columns]

## Problem

a) How many different states are there in the `dmv_highschool` dataset? Make sure you check before jump into a conclusion.

In [285]:
## your answer is here
unique_states=dmv_hs['state_name'].str.lower()
count=unique_states.nunique()
print('There are',count, 'different states in dmv_highschool dataset\n')
unique_states.unique()

There are 3 different states in dmv_highschool dataset



array(['virginia', 'maryland', 'district of columbia'], dtype=object)

b) Standardize the state names in __both__ dataframes. First letter capital, all others lower case. e.g. MArylanD --> Maryland. Double check your answer.

In [286]:
## your answer is here
maryland_hs['state_name']=maryland_hs['state_name'].str.capitalize()
dmv_hs['state_name']=dmv_hs['state_name'].str.capitalize()

In [287]:
maryland_hs['state_name']

0      Maryland
1      Maryland
2      Maryland
3      Maryland
4      Maryland
         ...   
199    Maryland
200    Maryland
201    Maryland
202    Maryland
203    Maryland
Name: state_name, Length: 204, dtype: object

In [288]:
dmv_hs['state_name']

0       Virginia
1       Virginia
2       Virginia
3       Virginia
4       Maryland
          ...   
3763    Maryland
3764    Maryland
3765    Virginia
3766    Virginia
3767    Virginia
Name: state_name, Length: 3768, dtype: object

In [289]:
## double check here
maryland_hs['state_name'].unique()

array(['Maryland'], dtype=object)

In [290]:
dmv_hs['state_name'].unique()

array(['Virginia', 'Maryland', 'District of columbia'], dtype=object)

## Problem

a) Check in `maryland_hs` dataset whether there are any repeating `school_names`

In [291]:
## your answer is here
school_names= pd.DataFrame(maryland_hs, columns=['school_name'])
dup_school_names = school_names[school_names.duplicated(['school_name'])]
print(dup_school_names)

                school_name
37          CHESAPEAKE HIGH
68  FREDERICK DOUGLASS HIGH


b) Check whether these schools are just a repetition or different schools.

In [292]:
## your answer is here
if maryland_hs.iloc[36].equals(maryland_hs.iloc[37]):
  print("\nIt's just a repeation")
else:
  print("\nBoth are different schools")  
if maryland_hs.iloc[67].equals(maryland_hs.iloc[68]):
  print("\nIt's just a repeation")
else:
  print("\nBoth are different schools")


Both are different schools

Both are different schools


c) Merge data frames in such a way for each high school on the `maryland_hs` we would have missing columns from `dmv_hs` dataset. Hint: Resulting dataframe should have shape (204, 21). If you cannot make it exactly in this format, submit the best solution you can find.

In [293]:
## your answer is here
merged_df=pd.merge(maryland_hs,dmv_hs)
merged_df.drop_duplicates('state_school_id',keep='first')

school_name  ... state_school_id
0                             ABERDEEN HIGH  ...    MD-12-120270
1        ACADEMY OF HEALTH SCIENCES AT PGCC  ...    MD-16-161350
2    ACHIEVEMENT ACADEMY @ HARBOR CITY HIGH  ...    MD-30-300413
3                      ALBERT EINSTEIN HIGH  ...    MD-15-150789
4                             ALLEGANY HIGH  ...    MD-01-010606
..                                      ...  ...             ...
201                         WILDE LAKE HIGH  ...    MD-13-130516
202                       WILLIAMSPORT HIGH  ...    MD-21-210204
203                  WINSTON CHURCHILL HIGH  ...    MD-15-150602
204                       WINTERS MILL HIGH  ...    MD-06-060704
205  WOODLAWN HIGH CENTER FOR PRE-ENG. RES.  ...    MD-03-030172

[204 rows x 21 columns]

## Problem

For each county find the average number of students. Sort the result in an descending order.


In [294]:
## your answer is here
merged_df.groupby('county_name')['total_students_all_grades'].mean().sort_values(ascending=False)

county_name
Montgomery County         1972.600000
Anne Arundel County       1761.230769
St. Mary's County         1719.000000
Howard County             1467.666667
Prince George's County    1352.148148
Calvert County            1279.750000
Charles County            1211.000000
Talbot County             1196.000000
Queen Anne's County       1189.000000
Baltimore County          1180.037037
Frederick County          1172.000000
Harford County            1124.500000
Wicomico County            955.750000
Cecil County               913.600000
Carroll County             911.111111
Washington County          850.285714
Caroline County            823.500000
Allegany County            742.666667
Baltimore city             704.346154
Worcester County           672.666667
Dorchester County          656.000000
Garrett County             577.000000
Kent County                563.000000
Name: total_students_all_grades, dtype: float64

## Problem

For each school find the ratio of `free_and_reduced_lunch_students` to the total number of students who are in the same __zipcode__ with this school. Record these values in a new column named `ratio`. Suppose in high school A there are 13 students eligile for free and reduced lunch and the zipcode of the school is 20001. If there is total of 200 students in the zipcode 20001 then for High School A your code should return 13/200. Hint: We used similar method at the end of sixth class.

In [295]:
## your answer is here
ratio_df=[]
i=0
for value in merged_df['free_and_reduced_lunch_students']:
  ratio= value/ merged_df[merged_df.location_zip==merged_df.location_zip[i]].total_students_all_grades.sum()
  i+=1
  ratio_df.append(ratio)
merged_df['ratio']=ratio_df
merged_df      

school_name  ...     ratio
0                             ABERDEEN HIGH  ...  0.453361
1        ACADEMY OF HEALTH SCIENCES AT PGCC  ...  0.066817
2    ACHIEVEMENT ACADEMY @ HARBOR CITY HIGH  ...  0.300219
3                      ALBERT EINSTEIN HIGH  ...  0.362543
4                             ALLEGANY HIGH  ...  0.183844
..                                      ...  ...       ...
201                         WILDE LAKE HIGH  ...  0.152773
202                       WILLIAMSPORT HIGH  ...  0.391982
203                  WINSTON CHURCHILL HIGH  ...  0.035474
204                       WINTERS MILL HIGH  ...  0.086626
205  WOODLAWN HIGH CENTER FOR PRE-ENG. RES.  ...  0.292805

[206 rows x 22 columns]

In [296]:
## Double Check for the zipcode 21157
merged_df[merged_df.location_zip == 21157]['free_and_reduced_lunch_students']/ merged_df[merged_df.location_zip == 21157].total_students_all_grades.sum()

62     0.002900
74     0.010149
197    0.061254
204    0.086626
Name: free_and_reduced_lunch_students, dtype: float64

In [297]:
## This cell should produce the same result as above.
merged_df[merged_df.location_zip == 21157].ratio

62     0.002900
74     0.010149
197    0.061254
204    0.086626
Name: ratio, dtype: float64

# Part-III: Submitting Your Work

[Please Submit Your Work to This Google Form](https://forms.gle/XgA7W56KuFAASYgJ9)